integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java

```diff
@@ -0,0 +1,279 @@
+  private void readTemplateFile() {
+    String fileLocation = BASE_TEMPLATE_PATH + getTemplateFile();
+
+    InputStream resource = getClass().getClassLoader().getResourceAsStream(fileLocation);
+
+    if (resource == null) {
+      LOGGER.error("Cannot read the template file {}. File not found.", fileLocation);
+      return;
+    }
+
+    try(BufferedReader reader = new BufferedReader(new InputStreamReader(resource))) {
+      String line;
+      StringBuilder responseData = new StringBuilder();
+
+      while ((line = reader.readLine()) != null) {
+        responseData.append(line);
+        responseData.append('\n');
+      }
+
+      this.messageMLTemplate = responseData.toString();
+    } catch (IOException e) {
+      LOGGER.error("Cannot read the template file " + fileLocation, e);
+    }
+  }
+
+  /**
+   * Generates MessageML V2 object according to the messageML template file and generated entity JSON
+   * parsing the JSON input payload.
+   *
+   * @param node JSON node received from the third-party service
+   * @return MessageML v2 object
+   */
+  public Message parse(JsonNode node) {
+    if (StringUtils.isEmpty(messageMLTemplate)) {
+      return null;
+    }
+
+    String entityJSON = getEntityJSON(node);
+
+    if (StringUtils.isNotEmpty(entityJSON)) {
+      Message message = new Message();
+      message.setMessage(messageMLTemplate);
+      message.setData(entityJSON);
+      message.setVersion(MessageMLVersion.V2);
+
+      return message;
+    }
+
+    return null;
+  }
+
+  /**
+   * Retrieves the Entity JSON based on metadata objects.
+   *
+   * @param node JSON input payload
+   * @return Entity JSON or null if there are no metadata objects to be processed.
+   */
+  private String getEntityJSON(JsonNode node) {
+    if (metadata == null) {
+      return null;
+    }
+
+    EntityObject root = new EntityObject(metadata.getType(), getVersion());
+    List<MetadataObject> objects = metadata.getObjects();
+
+    preProcessInputData(node);
+    processMetadataObjects(root, node, objects);
+    postProcessOutputData(root, node);
+
+    try {
+      Map<String, Object> result = new LinkedHashMap<>();
+      result.put(metadata.getName(), root);
+
+      return JsonUtils.writeValueAsString(result);
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Fail to parse incoming payload", e);
+      return null;
+    }
+  }
+
+  /**
+   * Perform a pre-processing on the input data.
+   *
+   * This method must be implemented by the concrete parser classes to perform any data manipulation
+   * on the JSON input payload or augment the payload received with internal Symphony data retrieved
+   * through the Symphony API's.
+   *
+   * Examples:
+   * - Replace '\n' to '<br/>' tags
+   * - Uppercase JSON field content
+   * - Scape characters
+   * - Include user ID and username retrieved from the User API
+   *
+   * @param input JSON input payload
+   */
+  protected abstract void preProcessInputData(JsonNode input);
```

Comment:
This doesn't need to be abstract.  You can provide an implementation that does nothing.  That way if subclass doesn't need to do any pre-processing, it doesn't need to implement a do-nothing method.